pkg/querier/frontend/frontend_querier_queues.go

```diff
@@ -0,0 +1,216 @@
+// If maxQueriers is <= 0, all queriers can handle this user's requests.
+// If maxQueriers has changed since the last call, queriers for this are recomputed.
+func (q *queues) getOrAddQueue(userID string, maxQueriers int) chan *request {
+	// Empty user is not allowed, as that would break our users list ("" is used for free spot).
+	if userID == "" {
+		return nil
+	}
+
+	if maxQueriers < 0 {
+		maxQueriers = 0
+	}
+
+	uq := q.userQueues[userID]
+
+	if uq == nil {
+		uq = &userQueue{
+			ch:    make(chan *request, q.maxUserQueueSize),
+			seed:  getSeedForUser(userID),
+			index: -1,
+		}
+		q.userQueues[userID] = uq
+
+		// Add user to the list of users... find first free spot, and put it there.
+		for ix, u := range q.users {
+			if u == "" {
+				uq.index = ix
+				q.users[ix] = userID
+				break
+			}
+		}
+
+		// ... or add to the end.
+		if uq.index < 0 {
+			uq.index = len(q.users)
+			q.users = append(q.users, userID)
+		}
+	}
+
+	if uq.maxQueriers != maxQueriers {
+		uq.maxQueriers = maxQueriers
+		uq.queriers = q.selectQueriersForUser(uq.seed, q.sortedQueriers, maxQueriers)
+	}
+
+	return uq.ch
+}
+
+// Finds next queue for the querier. To support fair scheduling between users, client is expected
+// to pass last user index returned by this function as argument. Is there was no previous
+// last user index, use -1.
+func (q *queues) getNextQueueForQuerier(lastUserIndex int, querier string) (chan *request, string, int) {
+	uid := lastUserIndex
+
+	for iters := 0; iters < len(q.users); iters++ {
+		uid = uid + 1
+
+		// Don't use "mod len(q.users)", as that could skip users at the beginning of the list
+		// for example when q.users has shrunk since last call.
+		if uid >= len(q.users) {
+			uid = 0
+		}
+
+		u := q.users[uid]
+		if u == "" {
+			continue
+		}
+
+		q := q.userQueues[u]
+
+		if q.queriers != nil {
+			if _, ok := q.queriers[querier]; !ok {
+				// This querier is not handling the user.
+				continue
+			}
+		}
+
+		return q.ch, u, uid
+	}
+	return nil, "", uid
+}
+
+func (q *queues) addQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+
+	q.querierConnections[querier] = conns + 1
+
+	// First connection from this querier.
+	if conns == 0 {
+		q.sortedQueriers = append(q.sortedQueriers, querier)
+		sort.Strings(q.sortedQueriers)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) removeQuerierConnection(querier string) {
+	conns := q.querierConnections[querier]
+	if conns <= 0 {
+		return
+	}
+
+	conns--
+	if conns > 0 {
+		q.querierConnections[querier] = conns
+	} else {
+		delete(q.querierConnections, querier)
+
+		ix := sort.SearchStrings(q.sortedQueriers, querier)
+		q.sortedQueriers = append(q.sortedQueriers[:ix], q.sortedQueriers[ix+1:]...)
+
+		q.recomputeUserQueriers()
+	}
+}
+
+func (q *queues) recomputeUserQueriers() {
+	for _, uq := range q.userQueues {
+		uq.queriers = q.selectQueriersForUser(uq.seed, q.sortedQueriers, uq.maxQueriers)
```

Comment:
should there be locking here?  I feel like this will be accessed from multiple goroutines.

Reply:
Adding / removing querier holds the lock in `*Frontend` already. All locking is done there, `*queues` doesn’t do any locking.